python/cugraph/cugraph/gnn/data_loading/bulk_sampler.py

```diff
@@ -0,0 +1,211 @@
+        output_path: str
+            The directory where results will be stored.
+        graph: cugraph.Graph
+            The cugraph graph to operate upon.
+        saturation_level: int (optional, default=200,000)
+            The number of samples that can be made within a single call.
+        batches_per_partition: int (optional, default=100)
+            The number of batches outputted to a single parquet partition.
+        rank: int (optional, default=0)
+            The rank of this sampler.  Used to isolate this sampler from
+            others that may be running on other nodes.
+        kwargs: kwargs
+            Keyword arguments to be passed to the sampler (i.e. fanout).
+        """
+        self.__batch_size = batch_size
+        self.__output_path = output_path
+        self.__graph = graph
+        self.__saturation_level = saturation_level
+        self.__batches_per_partition = batches_per_partition
+        self.__rank = rank
+        self.__batches = None
+        self.__sample_call_args = kwargs
+
+    @property
+    def rank(self) -> int:
+        return self.__rank
+
+    @property
+    def saturation_level(self) -> int:
+        return self.__saturation_level
+
+    @property
+    def batch_size(self) -> int:
+        return self.__batch_size
+
+    @property
+    def batches_per_partition(self) -> int:
+        return self.__batches_per_partition
+
+    @property
+    def size(self) -> int:
+        if self.__batches is None:
+            return 0
+        else:
+            return len(self.__batches)
+
+    def add_batches(
+        self,
+        df: Union[cudf.DataFrame, dask_cudf.DataFrame],
+        start_col_name: str,
+        batch_col_name: str,
+    ) -> None:
+        """
+        Adds batches to this BulkSampler.
+
+        Parameters
+        ----------
+        df: cudf.DataFrame or dask_cudf.DataFrame
+            Contains columns for vertex ids, batch id
+        start_col_name: str
+            Name of the column containing the start vertices
+        batch_col_name: str
+            Name of the column containing the batch ids
+
+        Returns
+        -------
+        """
+        df = df.rename(
+            columns={
+                start_col_name: self.start_col_name,
+                batch_col_name: self.batch_col_name,
+            }
+        )
+
+        if self.__batches is None:
+            self.__batches = df
+        else:
+            if isinstance(df, type(self.__batches)):
+                self.__batches = self.__batches.append(df)
+            else:
+                raise TypeError(
+                    "Provided batches must match the dataframe"
+                    " type of previous batches!"
+                )
+
+        if self.size >= self.saturation_level:
+            self.flush()
+
+    def flush(self, skip_partition_size_check=False) -> None:
+        """
+        Computes all uncomputed batches
+        """
+        if self.__batches is None:
```

Comment:
I had to do something like below to make it work otherwise it sometimes gets stuck in recursive problems but dont have a MRE .  Do you have a test for this @alexbarghi-nv  ?  
```suggestion
        if self.__batches is None or len(self.__batches) == 0:
               self.__batches = None
```

Reply:
I have not encountered this problem while testing the loader.

Reply:
but I have a slightly different fix for the bug

Reply:
@VibhuJawa this should be resolved after the latest push.

Reply:
So this [resolution](https://github.com/rapidsai/cugraph/pull/3148/commits/f51870e73e78131de32ff98a8003546f2045cc44) fails with 
```python
TypeError: object of type 'NoneType' has no len()
```
because  `self.__batches` can be None.

Reply:
Ok, please try again now.